src/zarr/buffer.py

```diff
@@ -436,6 +442,311 @@ def as_numpy_array_wrapper(
     return prototype.buffer.from_bytes(func(buf.as_numpy_array()))
 
 
+class GpuBuffer(Buffer):
+    """A flat contiguous memory block on the GPU
+
+    We use Buffer throughout Zarr to represent a contiguous block of memory.
+
+    A Buffer is backed by a underlying array-like instance that represents
+    the memory. The memory type is unspecified; can be regular host memory,
+    CUDA device memory, or something else. The only requirement is that the
+    array-like instance can be copied/converted to a regular Numpy array
+    (host memory).
+
+    Note
+    ----
+    This buffer is untyped, so all indexing and sizes are in bytes.
+
+    Parameters
+    ----------
+    array_like
+        array-like object that must be 1-dim, contiguous, and byte dtype.
+    """
+
+    def __init__(self, array_like: ArrayLike):
+        if cp is None:
+            raise ImportError("Cannot use GpuBuffer without cupy. Please install cupy.")
+
+        if array_like.ndim != 1:
+            raise ValueError("array_like: only 1-dim allowed")
+        if array_like.dtype != np.dtype("b"):
+            raise ValueError("array_like: only byte dtype allowed")
+
+        # if type(array_like) == cp.ndarray
+        #     self._data = array_like
+        if hasattr(array_like, "__cuda_array_interface__"):
+            self._data = cp.asarray(array_like)
+        else:
+            # Slow copy based path for arrays that don't support the __cuda_array_interface__
+            # TODO: Add a fast zero-copy path for arrays that support the dlpack protocol
+            msg = (
+                "Creating a GpuBuffer with an array that does not support the "
+                "__cuda_array_interface__ for zero-copy transfers, "
+                "falling back to slow copy based path"
+            )
+            warnings.warn(
+                msg,
+                stacklevel=2,
+            )
+            buffer = Buffer(array_like)
+            self._data = cp.asarray(buffer.as_numpy_array())
```

Comment:
Made this change. But we'd still need a separate path for DLPack right? Can probably deal with that in a later PR since cupy, torch and jax should all be covered by this.